### PR TITLE
SNOW-1571459 parse jbdc properties

### PR DESCRIPTION
**Pull request by sfc-gh-bzabek**
# Overview

[SNOW-1571459 parse properties](https://snowflakecomputing.atlassian.net/browse/SNOW-1571459)
[SNOW-1571515 detect overrides](https://snowflakecomputing.atlassian.net/browse/SNOW-1571515)

Add new property: `"snowflake.jdbc.map": "isInsecureMode : true, notYetExistingProp : true"`

This will allow user to set custom snowflake jdbc properties and to handle properties that will be created for JDBC Driver in the future.
The map is not allowed to override other jdbc properties like: `snowflake.database.name, snowflake.schema,name, snowflake.private.key` etc. nor `jvm.proxy.xxx`. In such a case the error is thrown.
It is also not designed to replace them.

Properties passed to a map are not validated, for example `"foo: bar"` is allowed. In such a case we rely on Jdbc Driver behaviour.

## Pre-review checklist
- [ ] This change should be part of a Behavior Change Release. See [go/behavior-change](http://go/behavior-change).
- [ ] This change has passed Merge gate tests
- [ ] Snowpipe Changes
- [ ] Snowpipe Streaming Changes
- [ ] This change is TEST-ONLY
- [ ] This change is README/Javadocs only
- [ ] This change is protected by a config parameter <PARAMETER_NAME> eg `snowflake.ingestion.method`.
    - [ ] `Yes` - Added end to end and Unit Tests. 
    - [ ] `No` - Suggest why it is not param protected
- [ ] Is his change protected by parameter <PARAMETER_NAME> on the server side?
    - [ ] The parameter/feature is not yet active in production (partial rollout or PrPr, see [Changes for Unreleased Features and Fixes](http://go/ppp-prpr)).
    - [ ] If there is an issue, it can be safely mitigated by turning the parameter off. This is also verified by a test (See [go/ppp](http://go/ppp)).

[SNOW-1571459]: https://snowflakecomputing.atlassian.net/browse/SNOW-1571459?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/JdbcProperties.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+import java.util.Properties;
+
+/** Wrapper class for all snowflake jdbc properties */
+public class JdbcProperties {
+
+  /** All jdbc properties including proxyProperties */
+  private final Properties properties;
+  /** Proxy related properties */
+  private final Properties proxyProperties;
```
Comment (sfc-gh-bzabek) [inline src/main/java/com/snowflake/kafka/connector/internal/JdbcProperties.java:11]: There is a special use case for it

Comment (sfc-gh-bzabek) [reply src/main/java/com/snowflake/kafka/connector/internal/JdbcProperties.java:11]:
```
    this.internalStage =
        new SnowflakeInternalStage(
            (SnowflakeConnectionV1) this.conn, credentialExpireTimeMillis, proxyProperties);
```

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java [hunk 2]
```diff
@@ -60,30 +58,27 @@ public class SnowflakeConnectionServiceV1 implements SnowflakeConnectionService
...
     this.stageType = null;
-    this.proxyProperties = proxyProperties;
     this.kafkaProvider = kafkaProvider;
+    Properties proxyProperties = jdbcProperties.getProxyProperties();
+    Properties combinedProperties = jdbcProperties.getProperties();
     try {
-      if (proxyProperties != null && !proxyProperties.isEmpty()) {
-        Properties combinedProperties =
-            mergeProxyAndConnectionProperties(this.prop, this.proxyProperties);
+      if (!proxyProperties.isEmpty()) {
```
Comment (sfc-gh-bzabek) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java:76]: nullability is handled in JdbcProperties create().

File: src/main/java/com/snowflake/kafka/connector/internal/JdbcProperties.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  private JdbcProperties(Properties combinedProperties, Properties proxyProperties) {
+    this.properties = combinedProperties;
+    this.proxyProperties = proxyProperties;
+  }
+
+  public Properties getProperties() {
+    Properties copy = new Properties();
```
Comment (sfc-gh-mbobowski) [inline src/main/java/com/snowflake/kafka/connector/internal/JdbcProperties.java:19]: Is there a reason for making a copy here?

Comment (sfc-gh-bzabek) [reply src/main/java/com/snowflake/kafka/connector/internal/JdbcProperties.java:19]: Super anti bug safety? I can remove it. I would not insist on keeping that.

File: src/test/java/com/snowflake/kafka/connector/internal/JdbcPropertiesTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+import static org.junit.Assert.*;
+
+import java.util.Properties;
+import org.junit.Test;
+
+public class JdbcPropertiesTest {
+
+  @Test
```
Comment (sfc-gh-mbobowski) [inline src/test/java/com/snowflake/kafka/connector/internal/JdbcPropertiesTest.java:11]: Please add new tests using JUnit5 (import from `junit.jupiter` package)

Comment (sfc-gh-bzabek) [reply src/test/java/com/snowflake/kafka/connector/internal/JdbcPropertiesTest.java:12]: Done in both tests.

File: src/test/java/com/snowflake/kafka/connector/internal/InternalUtilsTest.java [hunk 5]
```diff
@@ -134,4 +136,18 @@ public void testResultSize() throws SQLException {
     resultSet = new MockResultSetForSizeTest(100);
     assert InternalUtils.resultSize(resultSet) == 100;
   }
+
+  @Test
```
[the remark this answers is not in the document]
Comment (sfc-gh-bzabek) [reply src/test/java/com/snowflake/kafka/connector/internal/InternalUtilsTest.java:140]: Done, only `@Test` needed migration.

File: src/test/java/com/snowflake/kafka/connector/internal/JdbcPropertiesTest.java (hunk 4 above)
Comment (sfc-gh-bzabek) [reply src/test/java/com/snowflake/kafka/connector/internal/JdbcPropertiesTest.java:12]: Done in both classes.

File: src/main/java/com/snowflake/kafka/connector/internal/JdbcProperties.java (hunk 3 above)
Comment (sfc-gh-bzabek) [reply src/main/java/com/snowflake/kafka/connector/internal/JdbcProperties.java]: I reconsidered and removed the copy.

File: src/test/java/com/snowflake/kafka/connector/internal/JdbcPropertiesTest.java (hunk 4 above)
Comment (sfc-gh-mbobowski) [reply src/test/java/com/snowflake/kafka/connector/internal/JdbcPropertiesTest.java:12]: Thanks!

**Pull request merged**